source/game/system/KPadController.cc

```diff
@@ -147,6 +147,71 @@ void RaceInputState::reset() {
     trickRaw = 0;
 }
 
+/// @brief Checks if the input state is valid.
+/// @return If the input state is valid.
+bool RaceInputState::isValid() const {
+    if (!isButtonsValid()) {
+        return false;
+    }
+
+    if (!isStickValid(stick.x) || !isStickValid(stick.y)) {
+        return false;
+    }
+
+    if (!isTrickValid()) {
+        return false;
+    }
+
+    return true;
+}
+
+/// @brief Checks if there are any invalid buttons.
+/// @details Validatation with the previous input state doesn't happen because it doesn't exist.
+/// Therefore, we cannot check here if e.g. the drift button is pressed when it shouldn't be.
+/// @return If no invalid buttons are present.
+bool RaceInputState::isButtonsValid() const {
+    return !(buttons & ~0xf);
```

Comment:
Is this just checking to make sure the upper byte isn't set?

Reply:
The upper 12 bits, and yes.